maven-core/src/main/java/org/apache/maven/session/scope/internal/InstanceBuilder.java

```diff
@@ -0,0 +1,139 @@
+
+/**
+ * Builds a proxy instance which is backed by a scoped provider.
+ *
+ * @author Simon Taddiken
+ * @param <T> Type of the proxy to create.
+ */
+final class InstanceBuilder<T> {
+
+    private Class<T> superType;
+    private InvocationHandler dispatcher;
+
+    private InstanceBuilder(Class<T> superType) {
+        this.superType = superType;
+    }
+
+    /**
+     * Builds a scoped proxy instance which is a subtype of the given class.
+     *
+     * @param type The class.
+     * @return Builder object for further configuration.
+     */
+    public static <E> InstanceBuilder<E> forType(Class<E> type) {
+        Preconditions.checkNotNull(type, "type");
+        Preconditions.checkArgument(
+                !Modifier.isFinal(type.getModifiers()), "can not proxy final type %s", type.getName());
+        return new InstanceBuilder<>(type);
+    }
+
+    /**
+     * Specifies the scoped provider. Every method call on the object created by this
+     * builder will be delegated to the object returned by the given provider.
+     *
+     * @param provider The scoped provider.
+     * @return Builder object for further configuration.
+     */
+    public InstanceBuilder<T> dispatchTo(Provider<T> provider) {
+        Preconditions.checkNotNull(provider, "provider");
+        final InvocationHandler callback = (proxy, method, args) -> method.invoke(provider.get(), args);
+        this.dispatcher = callback;
+        return this;
+    }
+
+    /**
+     * Creates the scoped proxy object using the given provider.
+     *
+     * @return The scoped proxy object.
+     */
+    @SuppressWarnings("unchecked")
+    public T create() {
+        Preconditions.checkState(this.dispatcher != null, "no provider set");
+        Class<? extends T> enhanced = new ByteBuddy()
+                .subclass(superType)
+                .method(ElementMatchers.any())
+                .intercept(InvocationHandlerAdapter.of(this.dispatcher))
+                .make()
+                .load(superType.getClassLoader(), ClassLoadingStrategy.Default.INJECTION)
+                .getLoaded();
+        Errors errors = new Errors();
+        T proxyInstance = createInstanceWithNullValues(enhanced, errors);
+        errors.throwProvisionExceptionIfErrorsExist();
+        return proxyInstance;
+    }
+
+    private <T> T createInstanceWithNullValues(Class<T> proxyClass, Errors errors) {
+        try {
+            final Constructor<T> ctor = getConstructorFor(proxyClass);
+            final Object[] args = new Object[ctor.getParameterCount()];
+            return callConstructor(ctor, errors, args);
+        } catch (final NoSuchMethodException e) {
+            errors.addMessage(new Message(e.getMessage(), e));
+            return null;
+        }
+    }
+
+    private <T> Constructor<T> getConstructorFor(Class<T> proxyClass) throws NoSuchMethodException {
+        final Collection<Constructor<?>> ctors = Arrays.stream(proxyClass.getConstructors())
```

Comment:
Not sure if it is required, but from my testing I have the impression that one either need an `@Inject` annotated constructor or a default contructor without arguments.

Reply:
I ended up using objenesis in the latest commit to create the instance so that the requirements on the original class are lessened.